velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,73 +26,89 @@
-// TODO: Support multiple S3FileSystem instances using a cache
-static std::shared_ptr<S3FileSystem> s3fs = nullptr;
+/// Multiple S3 filesystems are supported.
+/// Key is the endpoint value specified in the config using hive.s3.endpoint.
+/// If the endpoint is empty, it will default to AWS S3.
+FileSystemMap& fileSystems() {
+  static FileSystemMap instances;
+  return instances;
+}
 
-std::function<std::shared_ptr<
-    FileSystem>(std::shared_ptr<const Config>, std::string_view)>
-fileSystemGenerator() {
-  static auto filesystemGenerator = [](std::shared_ptr<const Config> properties,
-                                       std::string_view filePath) {
-    folly::call_once(S3FSInstantiationFlag, [&properties]() {
-      std::shared_ptr<S3FileSystem> fs;
-      if (properties != nullptr) {
-        initializeS3(properties.get());
-        fs = std::make_shared<S3FileSystem>(properties);
-      } else {
-        auto config = std::make_shared<core::MemConfig>();
-        initializeS3(config.get());
-        fs = std::make_shared<S3FileSystem>(config);
-      }
-      s3fs = fs;
-    });
-    return s3fs;
-  };
-  return filesystemGenerator;
+std::string getS3Identity(std::shared_ptr<core::MemConfig>& config) {
+  HiveConfig hiveConfig = HiveConfig(config);
+  auto endpoint = hiveConfig.s3Endpoint();
+  if (!endpoint.empty()) {
+    // The identity is the endpoint.
+    return endpoint;
+  }
+  // Default key value.
+  return "aws-s3-key";
 }
 
-std::function<std::unique_ptr<velox::dwio::common::FileSink>(
-    const std::string&,
-    const velox::dwio::common::FileSink::Options& options)>
-s3WriteFileSinkGenerator() {
-  static auto s3WriteFileSink =
-      [](const std::string& fileURI,
-         const velox::dwio::common::FileSink::Options& options)
-      -> std::unique_ptr<dwio::common::WriteFileSink> {
-    if (isS3File(fileURI)) {
-      auto fileSystem =
-          filesystems::getFileSystem(fileURI, options.connectorProperties);
-      return std::make_unique<dwio::common::WriteFileSink>(
-          fileSystem->openFileForWrite(fileURI, {{}, options.pool}),
-          fileURI,
-          options.metricLogger,
-          options.stats);
-    }
-    return nullptr;
-  };
+std::shared_ptr<FileSystem> fileSystemGenerator(
+    std::shared_ptr<const Config> properties,
+    std::string_view /*filePath*/) {
+  std::shared_ptr<core::MemConfig> config = std::make_shared<core::MemConfig>();
+  if (properties) {
+    *config = core::MemConfig(properties->values());
+  }
+  const auto s3Identity = getS3Identity(config);
+
+  return fileSystems().withWLock(
+      [&](auto& instanceMap) -> std::shared_ptr<FileSystem> {
+        initializeS3(config.get());
+        if (instanceMap.count(s3Identity) == 0) {
+          auto fs = std::make_shared<S3FileSystem>(properties);
+          instanceMap.insert({s3Identity, fs});
+          return fs;
+        }
+        return instanceMap[s3Identity];
```

Comment:
should probably .find() instead of .count() so you don't need to look it up again here.

Reply:
fixed